File: server/vision-quality.mjs
// Deterministic extraction grading only: these scores never create financial facts.
export const visionGradeVersion=1;
export const visionMetrics=Object.freeze(['numeric','unit','date','header','sign','parentheses','footnote','relationship','missing']);
const keys=(value,expected)=>value&&typeof value==='object'&&!Array.isArray(value)&&Object.keys(value).sort().join('|')===[...expected].sort().join('|');
const label=value=>typeof value==='string'&&value.trim().length>0&&value.length<=500;
const coordinate=cell=>JSON.stringify([cell.row,cell.column]);
function decimal(value){
 if(typeof value!=='string'||value.length>100)return null;
 let token=value.trim().replaceAll('−','-'),parentheses=token.startsWith('(')&&token.endsWith(')');
 if(parentheses)token=token.slice(1,-1);
 const match=token.match(/^([+-]?)(\d{1,3}(?:,\d{3})+|\d+)(?:\.(\d+))?(%)?$/);
 if(!match||parentheses&&match[1])return null;
 const whole=BigInt(match[2].replaceAll(',','')).toString(),fraction=(match[3]??'').replace(/0+$/,'');
 return {magnitude:whole+(fraction?'.'+fraction:''),negative:parentheses||match[1]==='-',percent:!!match[4],parentheses};
}
function validTable(table){
 return keys(table,['headers','cells','footnotes'])&&Array.isArray(table.headers)&&table.headers.length>0&&table.headers.length<=30&&table.headers.every(label)&&
  Array.isArray(table.footnotes)&&table.footnotes.length<=30&&table.footnotes.every(label)&&
  Array.isArray(table.cells)&&table.cells.length>0&&table.cells.length<=200&&table.cells.every(cell=>keys(cell,['row','column','value','unit','date','footnote'])&&
   label(cell.row)&&label(cell.column)&&label(cell.unit)&&label(cell.date)&&(cell.footnote===null||label(cell.footnote))&&(cell.value===null||decimal(cell.value)!==null))&&
  new Set(table.cells.map(coordinate)).size===table.cells.length;
}
export function gradeVisionTable(expected,observed){
 if(!validTable(expected))throw new TypeError('Invalid Vision grading reference');
 let output=observed;
 if(typeof observed==='string')try{output=JSON.parse(observed.replace(/^```(?:json)?\s*/,'').replace(/\s*```$/,''));}catch{output=null;}
 const valid=validTable(output),metrics=Object.fromEntries(visionMetrics.map(name=>[name,{correct:0,total:expected.cells.length}]));
 let criticalErrors=0;
 const actual=new Map(valid?output.cells.map(cell=>[coordinate(cell),cell]):[]);
 const header=valid&&JSON.stringify(output.headers)===JSON.stringify(expected.headers);
 const footnotes=valid&&JSON.stringify(output.footnotes)===JSON.stringify(expected.footnotes);
 for(const cell of expected.cells){
  const found=actual.get(coordinate(cell)),want=decimal(cell.value),got=decimal(found?.value);
  const missing=!!found&&(cell.value===null)===(found.value===null);
  const numeric=!!found&&(cell.value===null?found.value===null:!!got&&got.magnitude===want.magnitude&&got.percent===want.percent);
  const sign=!!found&&(cell.value===null?found.value===null:!!got&&got.negative===want.negative);
  const parentheses=!!found&&(cell.value===null?found.value===null:!!got&&got.parentheses===want.parentheses);
  const unit=!!found&&found.unit===cell.unit,date=!!found&&found.date===cell.date,footnote=footnotes&&!!found&&found.footnote===cell.footnote;
  const relationship=numeric&&sign&&unit&&date&&footnote;
  const checks={numeric,unit,date,header,sign,parentheses,footnote,relationship,missing};
  for(const name of visionMetrics)if(checks[name])metrics[name].correct++;
  if(!Object.values(checks).every(Boolean))criticalErrors++;
 }
 if(valid)criticalErrors+=output.cells.filter(cell=>!expected.cells.some(want=>coordinate(want)===coordinate(cell))).length;
 for(const metric of Object.values(metrics))metric.score=metric.correct/metric.total;
 return {version:visionGradeVersion,formatValid:Boolean(valid),passed:Boolean(valid&&criticalErrors===0),criticalErrors,metrics,trust:'benchmark-only'};
}

File: python_backend/domain/vision_quality.py
"""Deterministic Vision table grading; results never become financial facts."""
from __future__ import annotations

import json
import re
from decimal import Decimal, InvalidOperation
from typing import Any, cast

VISION_GRADE_VERSION = 1
VISION_METRICS = ("numeric", "unit", "date", "header", "sign", "parentheses", "footnote", "relationship", "missing")


def _label(value: object) -> bool:
    return isinstance(value, str) and bool(value.strip()) and len(value) <= 500


def _decimal(value: object) -> dict[str, Any] | None:
    if not isinstance(value, str) or len(value) > 100:
        return None
    token = value.strip().replace("−", "-")
    parentheses = token.startswith("(") and token.endswith(")")
    if parentheses:
        token = token[1:-1]
    match = re.fullmatch(r"([+-]?)(\d{1,3}(?:,\d{3})+|\d+)(?:\.(\d+))?(%)?", token)
    if not match or parentheses and match.group(1):
        return None
    try:
        magnitude = Decimal((match.group(2) + ("." + match.group(3) if match.group(3) else "")).replace(",", ""))
    except InvalidOperation:
        return None
    return {
        "magnitude": magnitude.normalize(),
        "negative": parentheses or match.group(1) == "-",
        "percent": bool(match.group(4)),
        "parentheses": parentheses,
    }


def valid_table(value: object) -> bool:
    if not isinstance(value, dict) or set(value) != {"headers", "cells", "footnotes"}:
        return False
    headers, cells, footnotes = value["headers"], value["cells"], value["footnotes"]
    if not isinstance(headers, list) or not 1 <= len(headers) <= 30 or not all(_label(item) for item in headers):
        return False
    if not isinstance(footnotes, list) or len(footnotes) > 30 or not all(_label(item) for item in footnotes):
        return False
    if not isinstance(cells, list) or not 1 <= len(cells) <= 200:
        return False
    coordinates: set[tuple[str, str]] = set()
    for cell in cells:
        if not isinstance(cell, dict) or set(cell) != {"row", "column", "value", "unit", "date", "footnote"}:
            return False
        if not all(_label(cell[key]) for key in ("row", "column", "unit", "date")) or cell["footnote"] is not None and not _label(cell["footnote"]):
            return False
        if cell["value"] is not None and _decimal(cell["value"]) is None:
            return False
        coordinate = (cell["row"], cell["column"])
        if coordinate in coordinates:
            return False
        coordinates.add(coordinate)
    return True


def grade_vision_table(expected: dict[str, Any], observed: object) -> dict[str, Any]:
    if not valid_table(expected):
        raise TypeError("视觉质量参考表无效")
    output = observed
    if isinstance(observed, str):
        text = re.sub(r"^```(?:json)?\s*|\s*```$", "", observed.strip())
        try:
            output = json.loads(text)
        except json.JSONDecodeError:
            output = None
    valid = valid_table(output)
    parsed = cast(dict[str, Any], output) if valid else {"headers": [], "cells": [], "footnotes": []}
    expected_cells = expected["cells"]
    metrics: dict[str, dict[str, int | float]] = {name: {"correct": 0, "total": len(expected_cells)} for name in VISION_METRICS}
    actual = {(cell["row"], cell["column"]): cell for cell in parsed["cells"]} if valid else {}
    header = bool(valid and parsed["headers"] == expected["headers"])
    footnotes = bool(valid and parsed["footnotes"] == expected["footnotes"])
    critical_errors = 0
    for cell in expected_cells:
        found = actual.get((cell["row"], cell["column"]))
        want, got = _decimal(cell["value"]), _decimal(found["value"]) if found else None
        missing = bool(found is not None and (cell["value"] is None) == (found["value"] is None))
        numeric = bool(found is not None and (found["value"] is None if cell["value"] is None else got and want and got["magnitude"] == want["magnitude"] and got["percent"] == want["percent"]))
        sign = bool(found is not None and (found["value"] is None if cell["value"] is None else got and want and got["negative"] == want["negative"]))
        parentheses = bool(found is not None and (found["value"] is None if cell["value"] is None else got and want and got["parentheses"] == want["parentheses"]))
        unit = bool(found is not None and found["unit"] == cell["unit"])
        date = bool(found is not None and found["date"] == cell["date"])
        footnote = bool(footnotes and found is not None and found["footnote"] == cell["footnote"])
        relationship = numeric and sign and unit and date and footnote
        checks = {"numeric": numeric, "unit": unit, "date": date, "header": header, "sign": sign,
                  "parentheses": parentheses, "footnote": footnote, "relationship": relationship, "missing": missing}
        for name, passed in checks.items():
            metrics[name]["correct"] += int(passed)
        critical_errors += int(not all(checks.values()))
    if valid:
        expected_coordinates = {(cell["row"], cell["column"]) for cell in expected_cells}
        critical_errors += sum(coordinate not in expected_coordinates for coordinate in actual)
    for metric in metrics.values():
        metric["score"] = metric["correct"] / metric["total"]
    return {"version": VISION_GRADE_VERSION, "formatValid": valid, "passed": bool(valid and critical_errors == 0),
            "criticalErrors": critical_errors, "metrics": metrics, "trust": "benchmark-only"}
